jgrapht-core/src/main/java/org/jgrapht/alg/lca/EulerTourRMQLCAFinder.java

```diff
@@ -0,0 +1,227 @@
+
+/**
+ * Algorithm for computing lowest common ancestors in rooted trees and forests using the Euler tour approached
+ * introduced in <i>Berkman, Omer; Vishkin, Uzi (1993), "Recursive Star-Tree Parallel Data Structure",
+ * SIAM Journal on Computing, 22 (2): 221–242, doi:10.1137/0222017</i>
+ *
+ * Preprocessing Time complexity: $O(|V| log(|V|))
+ * Preprocessing Memory complexity:  $O(|V| log(|V|))
+ * Query complexity: $O(1)$
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Alexandru Valeanu
+ */
+public class EulerTourRMQLCAFinder<V, E> implements LCAAlgorithm<V> {
+    private final Graph<V, E> graph;
+    private final Set<V> roots;
+    private final int MAX_LEVEL;
+
+    private Map<V, Integer> vertexMap;
+    private List<V> indexList;
+
+    private int[] eulerTour;
+    private int sizeTour;
+
+    private int numberComponent;
+    private int[] component;
+
+    private int[] level;
+    private int[] representative;
+
+    private int[][] rmq;
+    private int[] log2;
+
+    /**
+     * Construct a new instance of the algorithm.
+     *
+     * Note: The constructor will NOT check if the input graph is a valid tree.
+     *
+     * @param graph the input graph
+     * @param root the root of the graph
+     */
+    public EulerTourRMQLCAFinder(Graph<V, E> graph, V root){
+        this(graph, Collections.singleton(Objects.requireNonNull(root, "root cannot be null")));
+    }
+
+    /**
+     * Construct a new instance of the algorithm.
+     *
+     * Note: If two roots appear in the same tree, an error will be thrown.
+     * Note: The constructor will NOT check if the input graph is a valid forest.
+     *
+     * @param graph the input graph
+     * @param roots the set of roots of the graph
+     */
+    public EulerTourRMQLCAFinder(Graph<V, E> graph, Set<V> roots){
+        this.graph = Objects.requireNonNull(graph, "graph cannot be null");
+        this.roots = Objects.requireNonNull(roots, "roots cannot be null");
+        this.MAX_LEVEL = 1 + log2(graph.vertexSet().size());
+
+        if (this.roots.isEmpty())
+            throw new IllegalArgumentException("roots cannot be empty");
+
+        if (!graph.vertexSet().containsAll(roots))
+            throw new IllegalArgumentException("at least one root is not a valid vertex");
+
+        computeAncestorsStructure();
+    }
+
+    private void normalizeGraph(){
+        /*
+         * Normalize the graph map each vertex to an integer (using a HashMap) keep the reverse
+         * mapping (using an ArrayList)
+         */
+        vertexMap = new HashMap<>(graph.vertexSet().size());
+        indexList = new ArrayList<>(graph.vertexSet().size());
+
+        for (V v : graph.vertexSet()) {
+            if (!vertexMap.containsKey(v)) {
+                vertexMap.put(v, vertexMap.size());
+                indexList.add(v);
+            }
+        }
+    }
+
+    private void dfs(int u, int parent, int lvl) {
```

Comment:
Could you convert this to iterative for better scalability?

Reply:
Done